docs/internal/versioning.md

```diff
@@ -0,0 +1,244 @@
+In Verona, visibility is regarded as purely a software-engineering tool for defining API-stability guarantees.
+As such, Verona provides only two levels of visibility:
+
+ - Public: may be accessed by anything
+ - Package-private: may be accessed only by code written in the same package
+
+These visibility rules apply to methods, fields, and types.
+They exist to establish a boundary for API-breaking changes.
+Adding or removing a private field or method does not constitute an API break.
+Changing the behaviour or a private method does not constitute an API break.
+Adding or removing a private type does not constitute an API break.
+All of these are true because no code outside of the package may name the types, fields, or methods that have changed.
+
+API-breaking changes
+--------------------
+
+Several kinds of change in Verona can introduce API-breaking changes.
+Note that we differentiate between *API breakage* and *implementation breakage*.
+Nothing that we do can prevent a buggy version of a package from being shipped.
+Even within the category of interface breakage, there are two key subcategories:
+
+ - Changes that affect things that are visible in the Verona type system and can cause a consumer of a package to fail to compile.
+ - Changes to behaviour documented via API contracts that can cause a consumer of a package to behave unexpectedly.
+
+For example, consider a map defines an iterator that returns the elements in insertion order.
+If a later version of the same package redefines this API to return the elements sorted by a total ordering on the keys, this is a different type.
+The API in both versions appears identical at the level of an abstract syntax tree but moving to the new version will break most consumers.
+We aim to provide tools to help developers avoid this kind of problem but we cannot (and do not try to) detect it automatically.
+
+There are a lot of ways of breaking a Verona API such that some consumers will fail to compile.
+These are [*Note* This should be an exhaustive list, but probably isn't.  We should add everything else we discover here]:
+
+ - Removing a public method or field from a class that is public.
+   This breaks any code that tries to call the method / access the field or that tries to cast a pointer to that concrete type to an interface type defining that member.
+ - Removing a public method field from a private class that is returned as an interface type from a public method.
+   Code outside the package may cast this to an interface defining the member and expose it.
+ - Adding a public method or field to a public class.
+   This will potentially allow the class to match interfaces that it did not previously and may affect the behaviour of code that assumes specific semantics from classes that match the interface.
+   It will also cause compile failures for any code that uses implementation inheritance to reuse the code from that class and defines methods or fields of the same name.
```

Comment:
If I understand the question correctly:

 - If you add a private field, an inheriting class from a different package doesn't see that field and so there's no conflict and because subclassing (implementation inheritance) does not define a subtyping relationship then there's no way that your class will ever be passed to something that expects the other type.  If you define an interface including the private field then the private field is actually defined by {method name, originating package} and so such an interface defined in the two packages will see different private fields.
 - If you add a public field, it's in the same namespace as fields of the inheriting class and so name conflicts can arrise.

I think this introduces a slight variation of the corner case below though.  Consider:

```verona
// Package A
class Base
{
  var x : U32;
  [[version=2]]
  var y : U64;
}

// Package B
import "A@1" as A;

class Sub : A.Base
{
  var y : U64;
}
```

If an instance of `Sub` is passed (as, for example an `Any`) to a third module that does import `A` buy defines an interface requiring a `var y : U64`, which does it see?  I think this one is probably the easiest to solve because there's no subtyping relationship between `Base` and `Sub`, so `Sub`'s definition defines which things are private and which things are public and in this case `Sub` interprets `Base`'s `y` as private and so does not expose it to *any* interface defined outside of package `B`.

Reply:
In our other (type inheritance) discussion, I thought the semantics was that a derived type cannot override a base's private field, even though it cannot see it. Like the `mixin` semantics, you type check the derived signature against the base class signature (that includes the private types).

With versioning, this becomes much harder, because API version 2 introduces new private fields (even if public, they're "private" to previous versions), but they're not really private, not in the class signature.

Does this mean API versioning will essentially change the type semantics? We'll need to carry version information around to the type checks to make sure we give meaningful error messages to users.